# Real-Time Bidding

## Problem 1 Data Exploration

In [1]:
import pandas as pd
Train_df = pd.read_csv('/Users/weisihan/Downloads/we_data/train.csv')


In [2]:
Train_df.describe()

click       weekday          hour        region          city  \
count  2.430981e+06  2.430981e+06  2.430981e+06  2.430981e+06  2.430981e+06   
mean   7.375623e-04  2.888556e+00  1.348925e+01  1.493833e+02  1.543295e+02   
std    2.714809e-02  2.055223e+00  6.567105e+00  1.020631e+02  1.033093e+02   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    0.000000e+00  1.000000e+00  1.000000e+01  7.900000e+01  7.900000e+01   
50%    0.000000e+00  3.000000e+00  1.400000e+01  1.460000e+02  1.520000e+02   
75%    0.000000e+00  5.000000e+00  1.900000e+01  2.160000e+02  2.220000e+02   
max    1.000000e+00  6.000000e+00  2.300000e+01  3.950000e+02  3.990000e+02   

         adexchange  urlid     slotwidth    slotheight     slotprice  \
count  2.381152e+06    0.0  2.430981e+06  2.430981e+06  2.430981e+06   
mean   2.109685e+00    NaN  5.774218e+02  1.885723e+02  2.671783e+01   
std    8.764370e-01    NaN  3.216738e+02  1.314946e+02  3.704444e+01   
min    1.000000e+00    NaN  1.200000e+02  5.000000e+01  0.000000e+00   
25%    1.000000e+00    NaN  3.000000e+02  9.000000e+01  0.000000e+00   
50%    2.000000e+00    NaN  3.360000e+02  9.000000e+01  5.000000e+00   
75%    3.000000e+00    NaN  9.600000e+02  2.500000e+02  5.000000e+01   
max    4.000000e+00    NaN  1.000000e+03  6.000000e+02  3.000000e+02   

           bidprice      payprice    advertiser  
count  2.430981e+06  2.430981e+06  2.430981e+06  
mean   2.729620e+02  7.815142e+01  2.840736e+03  
std    2.891153e+01  5.984636e+01  7.841990e+02  
min    2.270000e+02  0.000000e+00  1.458000e+03  
25%    2.410000e+02  3.300000e+01  2.259000e+03  
50%    2.940000e+02  6.800000e+01  3.358000e+03  
75%    3.000000e+02  9.900000e+01  3.427000e+03  
max    3.000000e+02  3.000000e+02  3.476000e+03

In [3]:
Train_df.head()

click  weekday  hour                                     bidid  \
0      0        5    22  b7bea80521fdecd95d2d761a38c91c3f09618066   
1      0        1    20  4f51205475678f5a124bc76b2c54163bf8eaa7eb   
2      0        3    13  b604e3fd054a658ab7ced4285ebf2ef54d2bd890   
3      0        6    23  0348beeae93e561584c3b50fc9e7746a33048ad7   
4      0        5     6  268149c1789bce2bc9798ffd97ec431219bafeb3   

                                     userid       useragent             IP  \
0  2e880fb7d690cf7377b2e42e701728e3f3c0e4c1      windows_ie   125.37.175.*   
1  3a1fe01360ff8100e7d006b83b77a3e4c01d928c  windows_chrome    171.36.92.*   
2  801d18a056b6fe6b06a794aef17fb0d6daff2414      windows_ie    59.46.106.*   
3  0d6eaf2259699990e38a1fc5116f112070b9ecdc      windows_ie  114.250.226.*   
4  a239d9bb642460d974ba67f85e63b8d3e214da0e      windows_ie   183.63.192.*   

   region  city  adexchange           ...            slotheight  \
0       2     2         2.0           ...                   200   
1     238   239         1.0           ...                   250   
2      40    41         2.0           ...                   250   
3       1     1         1.0           ...                   600   
4     216   233         2.0           ...                    90   

  slotvisibility  slotformat slotprice                          creative  \
0              2           0         5  a4f763f78ef3eedfe614263b94a8924e   
1     FourthView          Na         0                             10722   
2              2           0         5  798b2d49952d77f1eace9f23c210d0b5   
3              2           1         0  cb7c76e7784031272e37af8e7e9b062c   
4      OtherView          Na       133                              7330   

   bidprice payprice                           keypage  advertiser  \
0       238        5  0f951a030abdaedd733ee8d114ce2944        3427   
1       294       23                               NaN        2821   
2       238       24  0f951a030abdaedd733ee8d114ce2944        3427   
3       300       25  bebefa5efe83beee17a3d245e7c5085b        1458   
4       277      133                               NaN        2259   

                   usertag  
0                      NaN  
1                      NaN  
2  10052,10006,13866,10110  
3        13866,10063,10111  
4                      NaN  

[5 rows x 25 columns]

In [4]:
Train_df.count()

click             2430981
weekday           2430981
hour              2430981
bidid             2430981
userid            2430981
useragent         2430981
IP                2430981
region            2430981
city              2430981
adexchange        2381152
domain            2293846
url               2344169
urlid                   0
slotid            2430981
slotwidth         2430981
slotheight        2430981
slotvisibility    2430981
slotformat        2430981
slotprice         2430981
creative          2430981
bidprice          2430981
payprice          2430981
keypage           1925991
advertiser        2430981
usertag           1933502
dtype: int64

There are 2430981 pieces of impressions in total

### num Impressions

### Basic Statistics

In [5]:
Adv_group=Train_df.groupby(Train_df['advertiser'])
Clicks = Adv_group['click'].sum()
#Clicks
Imps = Adv_group['click'].count()
#Imps
Cost =  Adv_group['payprice'].sum()/1000
#Cost: the unit is fen/one impression, bidprice or payprice is fen oer 1000 impressions
CTR = Clicks/Imps
#CTR
CPM = Cost/Imps
CPM = CPM*1000
#CPM: cost per mille people, so *1000
eCPC = Cost/Clicks
#eCPC
Train_summ = pd.DataFrame({'Imps':Imps, 'Clicks':Clicks, 'Cost':Cost, 'CTR':CTR, 'CPM':CPM, 'eCPC':eCPC})
Train_summ

CPM       CTR  Clicks       Cost    Imps        eCPC
advertiser                                                            
1458        68.992646  0.000782     385  33968.736  492353   88.230483
2259        92.974931  0.000322      43  12428.238  133673  289.028791
2261        89.662184  0.000327      36   9873.779  110122  274.271639
2821        89.077922  0.000620     131  18828.044  211366  143.725527
2997        62.800116  0.004355     217   3129.267   49829   14.420585
3358        84.720599  0.000762     202  22447.231  264956  111.124906
3386        76.766320  0.000703     320  34931.823  455041  109.161947
3427        75.616329  0.000675     272  30458.711  402806  111.980555
3476        76.950083  0.000602     187  23918.779  310835  127.907909

In [6]:
Train_summ.columns

Index(['CPM', 'CTR', 'Clicks', 'Cost', 'Imps', 'eCPC'], dtype='object')

## Validation Set

In [1]:
!ls

Explore SherlockML	RandomForests_UCL      Unsupervised_Learning
Learn Data Science	script.ipynb	       validation.csv
__MACOSX		stage1_test.zip        Web Ind.ipynb
NeuralNets_UCL		stage1_train	       Welcome to SherlockML.ipynb
Optimisation (1).ipynb	stage1_train.zip
pandas-ucl		ucl_linear_regression


In [2]:
import pandas as pd
Validation_df = pd.read_csv('./validation.csv')


In [3]:
Validation_df['payprice'].sum()

23777270

In [4]:
Validation_df.groupby(Validation_df['advertiser'])['payprice'].sum()

advertiser
1458    4294602
2259    1568808
2261    1214876
2821    2394900
2997     388784
3358    2794021
3386    4350793
3427    3776735
3476    2993751
Name: payprice, dtype: int64

In [5]:
Validation_df.groupby(Validation_df['advertiser'])['payprice'].count()
Validation_df.head()

click  weekday  hour                                     bidid  \
0      0        4    20  bbcb813b6166538503d8b33a5602d7d72f6019dc   
1      0        1    21  5a07316c49477cb5d9b4d5aa39c27d6c3be7f92d   
2      0        4     8  f6ece71dae81d6b16bfb24ad6dd5611472d4c673   
3      0        5    15  b4d5c57c9b38ff5a12954fa01e11931b4e6bfbbb   
4      0        1    18  0899bf144249458ea9c89188473694bf44c7ca15   

                                     userid       useragent             IP  \
0  663169f66491c98c69f1f94a5c48fa34aa9fe06f      windows_ie  211.144.203.*   
1  a23f0cfab6592c137f796e68fa752fceba08bb1b  windows_chrome   58.247.250.*   
2  5105a013ea54a9706146033e6a138d5234a3803e      windows_ie   117.12.111.*   
3  577e9a768ce17e73c61ba7022db927f9ee761830  windows_chrome   61.187.224.*   
4  e4bf79c562745d671b19f2edf7fda89c2e25987f  windows_chrome   117.41.145.*   

   region  city  adexchange                       ...                         \
0      79    79         1.0                       ...                          
1      79    79         1.0                       ...                          
2       2     2         2.0                       ...                          
3     201   205         2.0                       ...                          
4     134   135         2.0                       ...                          

  slotheight slotvisibility  slotformat slotprice  \
0        600              2           1         0   
1         90              0           1         0   
2        250              2           0         5   
3        280              2           0         5   
4        200              1           0         5   

                           creative  bidprice payprice  \
0  cb7c76e7784031272e37af8e7e9b062c       300       23   
1  c938195f9e404b4f38c7e71bf50263e5       238       75   
2  911b2d84826786018761e8c0b0a3a60c       238       65   
3  dc0998c10f8f0b623b5d949e8272e4c7       238        6   
4  23485fcd23122d755d38f8c89d46ca56       249        5   

                            keypage  advertiser  \
0  bebefa5efe83beee17a3d245e7c5085b        1458   
1  a3becb967c015e9acb5e72f22b100ef8        3476   
2  d29e59bf0f7f8243858b8183f14d4412        3358   
3  d29e59bf0f7f8243858b8183f14d4412        3358   
4  7e0e909ea89f707bc20f9585c9072081        3476   

                                           usertag  
0                                      13866,10111  
1  10075,10057,10024,10052,10083,10063,10006,10110  
2  13866,10024,10059,10063,13776,10083,10006,10111  
3                    13866,10057,10006,10063,10110  
4                          10006,10063,13776,10110  

[5 rows x 25 columns]

In [6]:
#20.5642
Validation_df=Validation_df.sort_values(by="bidid" , ascending=True) 


In [7]:
Validation_df.head()

click  weekday  hour                                     bidid  \
176247      0        6    17  000007c951bc6bb61175fd16375abc655ba22438   
196052      0        5    14  00007d9e9c75eae1e33ca541088eb2cac3da6ac9   
171240      0        6    14  0000dea12a9b72a139ce072eecd4ee2fef52108f   
129174      0        4    23  00014f02887eda047f2260bc36e6aae16507a0df   
123696      0        6    17  00015c2145f816c225b7f1157d2398f0dbc0d393   

                                          userid       useragent  \
176247  b94636932735f79cee15ce0a9c2d015b239acb7a  windows_chrome   
196052  4ed990ed5e3591913d84a2e28c247631afed53e4      windows_ie   
171240  0c53cb61fe4f33809495eb49908db80666721eda      windows_ie   
129174  a3e3ec682b71e3c5768a6952480f4c12b87a4c9f      windows_ie   
123696  167fa4c56401a077ca5ca56755f1edff9e166b06      windows_ie   

                   IP  region  city  adexchange  \
176247    111.77.64.*     134   138         3.0   
196052    115.61.44.*     164   165         3.0   
171240     59.34.58.*     216   226         3.0   
129174    182.34.59.*     146   152         3.0   
123696  222.160.233.*      55    63         3.0   

                    ...               slotheight slotvisibility  slotformat  \
176247              ...                      250              0           0   
196052              ...                       90              0           0   
171240              ...                      250             Na          Na   
129174              ...                       90              0           0   
123696              ...                      250              0           0   

       slotprice                          creative  bidprice payprice  \
176247        31  e1af08818a6cd6bbba118bb54a651961       254       31   
196052        70  832b91d59d0cb5731431653204a76c0e       300       70   
171240        50                              7323       294      224   
129174        31  6cdf8fdd3e01122b09b5b411510a2385       300      120   
123696        70  00fccc64a1ee2809348509b7ac2a97a5       241       77   

                                 keypage  advertiser  \
176247  a5d0bde104fea52ea466c77b37948424        3476   
196052  bebefa5efe83beee17a3d245e7c5085b        1458   
171240                               NaN        2259   
129174  43f4a37f42a7c5e6219e2601b26c6976        3386   
123696  b2e35064f3549d447edbbdfb1f707c8c        3427   

                              usertag  
176247  10075,10063,10006,10083,10110  
196052                    10083,10110  
171240                            NaN  
129174  10063,13866,10031,10006,10110  
123696        10063,13403,10006,10057  

[5 rows x 25 columns]

In [8]:
Validation_df['bidprice'] = 23

In [9]:
def Prob2 (bidprice):
    if budget>0:
        Validation_df['win'] = Validation_df.apply(lambda row: valuation_formula(row['bidprice'], row['payprice']), axis=1)
    
    
    
    

In [10]:
def valuation_formula(x, y):
    if x>y:
        return 1
    else:
        return 0



In [11]:
Validation_df.head()

click  weekday  hour                                     bidid  \
176247      0        6    17  000007c951bc6bb61175fd16375abc655ba22438   
196052      0        5    14  00007d9e9c75eae1e33ca541088eb2cac3da6ac9   
171240      0        6    14  0000dea12a9b72a139ce072eecd4ee2fef52108f   
129174      0        4    23  00014f02887eda047f2260bc36e6aae16507a0df   
123696      0        6    17  00015c2145f816c225b7f1157d2398f0dbc0d393   

                                          userid       useragent  \
176247  b94636932735f79cee15ce0a9c2d015b239acb7a  windows_chrome   
196052  4ed990ed5e3591913d84a2e28c247631afed53e4      windows_ie   
171240  0c53cb61fe4f33809495eb49908db80666721eda      windows_ie   
129174  a3e3ec682b71e3c5768a6952480f4c12b87a4c9f      windows_ie   
123696  167fa4c56401a077ca5ca56755f1edff9e166b06      windows_ie   

                   IP  region  city  adexchange  \
176247    111.77.64.*     134   138         3.0   
196052    115.61.44.*     164   165         3.0   
171240     59.34.58.*     216   226         3.0   
129174    182.34.59.*     146   152         3.0   
123696  222.160.233.*      55    63         3.0   

                    ...               slotheight slotvisibility  slotformat  \
176247              ...                      250              0           0   
196052              ...                       90              0           0   
171240              ...                      250             Na          Na   
129174              ...                       90              0           0   
123696              ...                      250              0           0   

       slotprice                          creative  bidprice payprice  \
176247        31  e1af08818a6cd6bbba118bb54a651961        23       31   
196052        70  832b91d59d0cb5731431653204a76c0e        23       70   
171240        50                              7323        23      224   
129174        31  6cdf8fdd3e01122b09b5b411510a2385        23      120   
123696        70  00fccc64a1ee2809348509b7ac2a97a5        23       77   

                                 keypage  advertiser  \
176247  a5d0bde104fea52ea466c77b37948424        3476   
196052  bebefa5efe83beee17a3d245e7c5085b        1458   
171240                               NaN        2259   
129174  43f4a37f42a7c5e6219e2601b26c6976        3386   
123696  b2e35064f3549d447edbbdfb1f707c8c        3427   

                              usertag  
176247  10075,10063,10006,10083,10110  
196052                    10083,10110  
171240                            NaN  
129174  10063,13866,10031,10006,10110  
123696        10063,13403,10006,10057  

[5 rows x 25 columns]

In [12]:
clc=Validation_df.groupby(Validation_df['advertiser'])['win'].sum()
imp=Validation_df.groupby(Validation_df['advertiser'])['win'].count()

KeyError: 'Column not found: win'

In [13]:
clc/imp

NameError: name 'clc' is not defined

In [14]:
from collections import namedtuple

def myiter(d, cols=None):
    if cols is None:
        v = d.values.tolist()
        cols = d.columns.values.tolist()
    else:
        j = [d.columns.get_loc(c) for c in cols]
        v = d.values[:, j].tolist()

    n = namedtuple('MyTuple', cols)

    for line in iter(v):
        yield n(*line)


In [30]:
Validation_df["win"] = 0

In [41]:
budget=6250000
for i in Validation_df.index:
    if budget>0:
        if Validation_df.at[i, 'bidprice']>Validation_df.at[i, 'payprice']:
            Validation_df.at[i, 'win'] = int(1)
        else:
            Validation_df.at[i, 'win'] = int(0)
        budget=budget-23
    else:
        break

In [42]:
budget

-20

In [43]:
Validation_df.head(100)

click  weekday  hour                                     bidid  \
176247      0        6    17  000007c951bc6bb61175fd16375abc655ba22438   
196052      0        5    14  00007d9e9c75eae1e33ca541088eb2cac3da6ac9   
171240      0        6    14  0000dea12a9b72a139ce072eecd4ee2fef52108f   
129174      0        4    23  00014f02887eda047f2260bc36e6aae16507a0df   
123696      0        6    17  00015c2145f816c225b7f1157d2398f0dbc0d393   
59535       0        5    10  00016062d5f2b78f1f54def9652602fea14d43fd   
286585      0        5     0  000262db1539e1003802e719797b855f07b816c3   
141319      0        0     0  000279f4cf4dc4921c6053860d8c2b44f5dd64fd   
205646      0        2     0  000294397eccea2bc4d55806b992575a8b51b9af   
271801      0        6     7  0002b962ffca99bdea3394ede434e572e9c375a5   
97877       0        2    22  0002c0654eec175c2b5c86b2e765fe4a4af99caa   
22697       0        4     8  0002d6dd50d21bce65724cf3297d8cb6175218ff   
189473      0        3    22  000302f67e011e8cff2ed5ecb07c4814a334269f   
111389      0        3    22  00030f5363e1f648c863dcda53ba7030fa103b30   
162717      0        4    21  00037ae7f660017d78984a9583b076f6270057b3   
151408      0        3    11  000385557a23ece29bf80ad4e15f700d62360579   
285457      0        6    21  00039941aca66b10cc914b66879af548a381aaf5   
289425      0        2    23  000399c13ad9caa9de529d2050b64f0cf999ad40   
21397       0        3    17  0003c0be240569a16c28fa9cddf913c2deaa92e3   
76889       0        2    16  0003f4aa8b71ee4e1c31d53f9bfc81cc85b45f28   
21572       0        5     1  0004865c35fcbd46cc3bcbf41be8c210c7b31491   
22816       0        4     1  0004c44c3ac4586c40d26ac2d0b28d801e6dbe96   
203080      0        1    15  00050d6389761115bc865ab2b4e121069da6aa0b   
117347      0        1    15  00051c59217a0b0c43dec1e098aa578f94eb6b6f   
238180      0        5     9  0005507f6e2ef7f438b9a5b8ef3410d32a884d61   
24394       0        3    22  00057f414c7f1e629644ff29a06e46289f9b1862   
46342       0        6     6  0005e24158a424395f982cabf1d2a2d19b2ff80c   
8223        0        6    23  000602034e686863a1fddbe3f21b679a66a6bf5d   
14985       0        6    17  000617d799815b9c6897e19f1a28ba81ca951594   
273332      0        1    19  00066776df0cb5b055fe37bc49f288ce62afe246   
...       ...      ...   ...                                       ...   
178476      0        0    20  00158234064d2937fef8a729a57e3b67292b7a4c   
17414       0        3    18  0015dfffbbde82908a31f40e8532b3bcb76ab0c9   
93861       0        1    23  00166e2348363dc74eb943be884e37edd8f675bc   
136670      0        1     8  0016d086c355850574b241a6f1cbc79386032a70   
206462      0        3    23  00174bc3baf876d26e6b867480ca4aad4be9d451   
209826      0        3    12  00175d67560ff274167bf5bf3395550f761a8a1c   
282054      0        3     0  00177007f8231fd99d07820e64223554ec6672d6   
121973      0        0    17  0018bfde4866d8347c68c8aca0522778716d35f0   
110609      0        5    12  0018c09b7ab40f27998adbe2f52a3368a37e6ccd   
219204      0        0     0  0018c7f502a4c3067d70497249d63cbe3bcecc9d   
277613      0        3    20  00193dbfd28bf9e444dbee3397712365416be481   
73114       0        0    14  00195761ce0af4f2d82d75136effc47378506c5b   
12969       0        2    20  00197075961492c27d2c5f8eb4a3fd82b11ec035   
287105      0        5    18  001972fa01cbd7c24ccc66a3b80ced39a30de964   
208004      0        3    13  001979c693ac8deb2755ef484eb1665326abe110   
250817      0        5    10  0019a1f0362e796c8d2817149708af288e8dd4c7   
256907      0        1    19  0019eba6efa2ecc93da3858d1ed3bb9a64167376   
233953      0        3    13  001a33c54a7c4c01f9559d9b2b8fa059cb6bbcda   
279964      0        2    17  001a3cf49e386270050b191e7ab7625cc128e61a   
292971      0        1    10  001ad49ce74c114b90cfe72592f41de28e4f0226   
39657       0        1    19  001af856dd6109e6468565ab553a8e3e59ae933f   
137524      0        5     0  001b34f1542fa2bf434d081d7c7c2f7c1289afd3   
57892       

In [ ]:
advertiser-23
1458    0.195756
2259    0.192761
2261    0.262657
2821    0.152737
2997    0.292260
3358    0.134491
3386    0.202277
3427    0.167387
3476    0.164131

advertiser-20.56
1458    0.182814
2259    0.181155
2261    0.251144
2821    0.143342
2997    0.272830
3358    0.110477
3386    0.195818
3427    0.103541
3476    0.100847

advertiser-18
1458    0.117284
2259    0.107688
2261    0.141402
2821    0.093272
2997    0.240771
3358    0.086038
3386    0.080279
3427    0.082239
3476    0.080714

In [44]:
def Prob1(consp_arr):
    Validation_df['win']=0
    Validation_df['bidprice'] = consp_arr
    budget=6250000
    for i in Validation_df.index:
        if budget>0:
            if Validation_df.at[i, 'bidprice']>Validation_df.at[i, 'payprice']:
                Validation_df.at[i, 'win'] = int(1)
            else:
                Validation_df.at[i, 'win'] = int(0)
            budget = budget - consp_arr
        else:
            break
            

            
    Validation_df_new = Validation_df[Validation_df['win']!=0]
    print(Validation_df_new.shape)
    clc = Validation_df_new.groupby(Validation_df_new['advertiser'])['click'].sum()
    imp = Validation_df_new.groupby(Validation_df_new['advertiser'])['click'].count()
    ser = clc/imp
    #vali_ctr_suem = pd.DataFrame({'ser':ser})
    print(ser)
    return

In [45]:
Prob1(23)

(49609, 26)
advertiser
1458    0.000549
2259    0.000000
2261    0.000000
2821    0.000000
2997    0.001859
3358    0.000000
3386    0.000097
3427    0.000266
3476    0.000000
Name: click, dtype: float64


In [46]:
Prob1(5)

(2176, 26)
advertiser
1458    0.0
2259    0.0
2261    0.0
2821    0.0
2997    0.0
3358    0.0
3386    0.0
3427    0.0
3476    0.0
Name: click, dtype: float64
